File: sdn-js/src/subscription.test.ts
/**
 * Tests for the SDN Subscription module
 */

import { describe, it, expect, beforeEach } from 'vitest';
import {
  SubscriptionManager,
  SubscriptionConfig,
  QueryFilter,
  RoutingHeader,
  evaluateFilter,
  evaluateFilters,
  validateSubscriptionConfig,
  createDefaultConfig,
  generateSubscriptionId,
  serializeRoutingHeader,
  deserializeRoutingHeader,
  getSchemaRoutingTopic,
  getPeerRoutingTopic,
} from './subscription';

describe('SubscriptionManager', () => {
  let manager: SubscriptionManager;

  beforeEach(() => {
    manager = new SubscriptionManager();
  });

  describe('createSubscription', () => {
    it('should create a valid subscription', () => {
      const config: SubscriptionConfig = {
        dataTypes: ['OMM.fbs', 'CDM.fbs'],
        sourcePeers: ['all'],
        encrypted: true,
        streaming: true,
      };

      const sub = manager.createSubscription(config);

      expect(sub.id).toMatch(/^sub_/);
      expect(sub.status).toBe('active');
      expect(sub.config.dataTypes).toEqual(['OMM.fbs', 'CDM.fbs']);
      expect(sub.messageCount).toBe(0);
    });

    it('should throw error for empty data types', () => {
      const config: SubscriptionConfig = {
        dataTypes: [],
        sourcePeers: ['all'],
        encrypted: true,
        streaming: true,
      };

      expect(() => manager.createSubscription(config)).toThrow('At least one data type');
    });

    it('should throw error for invalid schema', () => {
      const config: SubscriptionConfig = {
        dataTypes: ['INVALID.fbs'],
        sourcePeers: ['all'],
        encrypted: true,
        streaming: true,
      };

      expect(() => manager.createSubscription(config)).toThrow('Unknown data type');
    });
  });

  describe('subscription lifecycle', () => {
    it('should pause and resume subscriptions', () => {
      const config: SubscriptionConfig = {
        dataTypes: ['OMM.fbs'],
        sourcePeers: ['all'],
        encrypted: true,
        streaming: true,
      };

      const sub = manager.createSubscription(config);
      expect(sub.status).toBe('active');

      manager.pauseSubscription(sub.id);
      expect(manager.getSubscription(sub.id)?.status).toBe('paused');

      manager.resumeSubscription(sub.id);
      expect(manager.getSubscription(sub.id)?.status).toBe('active');
    });

    it('should remove subscriptions', () => {
      const config: SubscriptionConfig = {
        dataTypes: ['OMM.fbs'],
        sourcePeers: ['all'],
        encrypted: true,
        streaming: true,
      };

      const sub = manager.createSubscription(config);
      expect(manager.getSubscription(sub.id)).toBeDefined();

      manager.removeSubscription(sub.id);
      expect(manager.getSubscription(sub.id)).toBeUndefined();
    });
  });

  describe('processMessage', () => {
    it('should deliver matching messages', () => {
      const config: SubscriptionConfig = {
        dataTypes: ['OMM.fbs'],
        sourcePeers: ['all'],
        encrypted: true,
        streaming: true,
      };

      const sub = manager.createSubscription(config);
      let receivedData: unknown = null;

      manager.addEventListener(sub.id, (event) => {
        if (event.type === 'message') {
          receivedData = event.data;
        }
      });

      const header: RoutingHeader = {
        schemaType: 'OMM',
        destinationPeers: [],
        ttl: 7,
        priority: 64,
        encrypted: true,
      };

      manager.processMessage('OMM.fbs', { OBJECT_NAME: 'ISS' }, 'peer1', header);

      expect(receivedData).toEqual({ OBJECT_NAME: 'ISS' });
      expect(manager.getSubscription(sub.id)?.messageCount).toBe(1);
    });

    it('should not deliver non-matching schema', () => {
      const config: SubscriptionConfig = {
        dataTypes: ['OMM.fbs'],
        sourcePeers: ['all'],
        encrypted: true,
        streaming: true,
      };

      const sub = manager.createSubscription(config);
      let received = false;

      manager.addEventListener(sub.id, () => {
        received = true;
      });

      manager.processMessage('CDM.fbs', {}, 'peer1');

      expect(received).toBe(false);
    });

    it('should filter by source peer', () => {
      const config: SubscriptionConfig = {
        dataTypes: ['OMM.fbs'],
        sourcePeers: ['peer1', 'peer2'],
        encrypted: true,
        streaming: true,
      };

      const sub = manager.createSubscription(config);
      const receivedFrom: string[] = [];

      manager.addEventListener(sub.id, (event) => {
        if (event.type === 'message' && event.from) {
          receivedFrom.push(event.from);
        }
      });

      manager.processMessage('OMM.fbs', {}, 'peer1');
      manager.processMessage('OMM.fbs', {}, 'peer2');
      manager.processMessage('OMM.fbs', {}, 'peer3'); // Should not match

      expect(receivedFrom).toEqual(['peer1', 'peer2']);
    });

    it('should apply field-level filters', () => {
      const config: SubscriptionConfig = {
        dataTypes: ['OMM.fbs'],
        sourcePeers: ['all'],
        encrypted: true,
        streaming: true,
        filters: [
          { field: 'OBJECT_NAME', operator: 'eq', value: 'ISS' },
        ],
      };

      const sub = manager.createSubscription(config);
      let matchCount = 0;

      manager.addEventListener(sub.id, (event) => {
        if (event.type === 'message') {
          matchCount++;
        }
      });

      manager.processMessage('OMM.fbs', { OBJECT_NAME: 'ISS' }, 'peer1');
      manager.processMessage('OMM.fbs', { OBJECT_NAME: 'Hubble' }, 'peer1');

      expect(matchCount).toBe(1);
    });
  });

  describe('getRequiredTopics', () => {
    it('should return schema and peer topics', () => {
      manager.createSubscription({
        dataTypes: ['OMM.fbs', 'CDM.fbs'],
        sourcePeers: ['all'],
        encrypted: true,
        streaming: true,
      });

      manager.createSubscription({
        dataTypes: ['EPM.fbs'],
        sourcePeers: ['peer1'],
        encrypted: true,
        streaming: true,
      });

      const topics = manager.getRequiredTopics();

      expect(topics.has('/sdn/data/OMM')).toBe(true);
      expect(topics.has('/sdn/data/CDM')).toBe(true);
      expect(topics.has('/sdn/data/EPM')).toBe(true);
      expect(topics.has('/sdn/peer/peer1')).toBe(true);
    });
  });
});

describe('evaluateFilter', () => {
  it('should evaluate eq operator', () => {
    const data = { name: 'ISS' };
    expect(evaluateFilter(data, { field: 'name', operator: 'eq', value: 'ISS' })).toBe(true);
    expect(evaluateFilter(data, { field: 'name', operator: 'eq', value: 'Hubble' })).toBe(false);
  });

  it('should evaluate ne operator', () => {
    const data = { name: 'ISS' };
    expect(evaluateFilter(data, { field: 'name', operator: 'ne', value: 'Hubble' })).toBe(true);
    expect(evaluateFilter(data, { field: 'name', operator: 'ne', value: 'ISS' })).toBe(false);
  });

  it('should evaluate numeric operators', () => {
    const data = { altitude: 400 };
    expect(evaluateFilter(data, { field: 'altitude', operator: 'gt', value: 300 })).toBe(true);
    expect(evaluateFilter(data, { field: 'altitude', operator: 'gte', value: 400 })).toBe(true);
    expect(evaluateFilter(data, { field: 'altitude', operator: 'lt', value: 500 })).toBe(true);
    expect(evaluateFilter(data, { field: 'altitude', operator: 'lte', value: 400 })).toBe(true);
  });

  it('should evaluate string operators', () => {
    const data = { description: 'International Space Station' };
    expect(evaluateFilter(data, { field: 'description', operator: 'contains', value: 'Space' })).toBe(true);
    expect(evaluateFilter(data, { field: 'description', operator: 'startsWith', value: 'International' })).toBe(true);
    expect(evaluateFilter(data, { field: 'description', operator: 'endsWith', value: 'Station' })).toBe(true);
  });

  it('should evaluate in/notIn operators', () => {
    const data = { type: 'satellite' };
    expect(evaluateFilter(data, { field: 'type', operator: 'in', value: ['satellite', 'debris'] })).toBe(true);
    expect(evaluateFilter(data, { field: 'type', operator: 'notIn', value: ['debris', 'rocket'] })).toBe(true);
  });

  it('should handle nested fields', () => {
    const data = { object: { name: 'ISS', orbit: { altitude: 400 } } };
    expect(evaluateFilter(data, { field: 'object.name', operator: 'eq', value: 'ISS' })).toBe(true);
    expect(evaluateFilter(data, { field: 'object.orbit.altitude', operator: 'gt', value: 300 })).toBe(true);
  });
});

describe('evaluateFilters', () => {
  it('should return true for empty filters', () => {
    expect(evaluateFilters({}, [])).toBe(true);
  });

  it('should require all filters to match (AND logic)', () => {
    const data = { name: 'ISS', altitude: 400 };
    const filters: QueryFilter[] = [
      { field: 'name', operator: 'eq', value: 'ISS' },
      { field: 'altitude', operator: 'gt', value: 300 },
    ];
    expect(evaluateFilters(data, filters)).toBe(true);

    const failingFilters: QueryFilter[] = [
      { field: 'name', operator: 'eq', value: 'ISS' },
      { field: 'altitude', operator: 'lt', value: 300 }, // This fails
    ];
    expect(evaluateFilters(data, failingFilters)).toBe(false);
  });
});

describe('validateSubscriptionConfig', () => {
  it('should return no errors for valid config', () => {
    const config: SubscriptionConfig = {
      dataTypes: ['OMM.fbs'],
      sourcePeers: ['all'],
      encrypted: true,
      streaming: true,
    };
    expect(validateSubscriptionConfig(config)).toEqual([]);
  });

  it('should return error for empty dataTypes', () => {
    const config: SubscriptionConfig = {
      dataTypes: [],
      sourcePeers: ['all'],
      encrypted: true,
      streaming: true,
    };
    const errors = validateSubscriptionConfig(config);
    expect(errors.some(e => e.includes('data type'))).toBe(true);
  });

  it('should return error for invalid operator in filter', () => {
    const config: SubscriptionConfig = {
      dataTypes: ['OMM.fbs'],
      sourcePeers: ['all'],
      encrypted: true,
      streaming: true,
      filters: [{ field: 'test', operator: 'invalid' as any, value: 'x' }],
    };
    const errors = validateSubscriptionConfig(config);
    expect(errors.some(e => e.includes('invalid operator'))).toBe(true);
  });
});

describe('createDefaultConfig', () => {
  it('should create config with defaults', () => {
    const config = createDefaultConfig();
    expect(config.dataTypes).toEqual([]);
    expect(config.sourcePeers).toEqual(['all']);
    expect(config.encrypted).toBe(true);
    expect(config.streaming).toBe(true);
    expect(config.rateLimit).toBe(1000);
  });
});

describe('generateSubscriptionId', () => {
  it('should generate unique IDs', () => {
    const id1 = generateSubscriptionId();
    const id2 = generateSubscriptionId();
    expect(id1).not.toBe(id2);
    expect(id1).toMatch(/^sub_/);
    expect(id2).toMatch(/^sub_/);
  });
});

describe('RoutingHeader serialization', () => {
  it('should serialize and deserialize minimal header', () => {
    const header: RoutingHeader = {
      schemaType: 'OMM',
      destinationPeers: [],
      ttl: 7,
      priority: 64,
      encrypted: true,
    };

    const serialized = serializeRoutingHeader(header);
    const deserialized = deserializeRoutingHeader(serialized);

    expect(deserialized).not.toBeNull();
    expect(deserialized?.schemaType).toBe('OMM');
    expect(deserialized?.ttl).toBe(7);
    expect(deserialized?.priority).toBe(64);
    expect(deserialized?.encrypted).toBe(true);
  });

  it('should serialize and deserialize header with destinations', () => {
    const header: RoutingHeader = {
      schemaType: 'CDM',
      destinationPeers: ['peer1', 'peer2', 'peer3'],
      ttl: 5,
      priority: 128,
      encrypted: false,
    };

    const serialized = serializeRoutingHeader(header);
    const deserialized = deserializeRoutingHeader(serialized);

    expect(deserialized).not.toBeNull();
    expect(deserialized?.schemaType).toBe('CDM');
    expect(deserialized?.destinationPeers).toEqual(['peer1', 'peer2', 'peer3']);
    expect(deserialized?.ttl).toBe(5);
    expect(deserialized?.encrypted).toBe(false);
  });

  it('should serialize and deserialize header with session key', () => {
    const header: RoutingHeader = {
      schemaType: 'EPM',
      destinationPeers: [],
      ttl: 7,
      priority: 64,
      encrypted: true,
      sessionKeyId: 'session123',
    };

    const serialized = serializeRoutingHeader(header);
    const deserialized = deserializeRoutingHeader(serialized);

    expect(deserialized?.sessionKeyId).toBe('session123');
  });

  it('should return null for invalid data', () => {
    expect(deserializeRoutingHeader(new Uint8Array([]))).toBeNull();
    expect(deserializeRoutingHeader(new Uint8Array([0x03, 0x4f, 0x4d]))).toBeNull();
  });
});

describe('topic helpers', () => {
  it('should generate schema routing topic', () => {
    expect(getSchemaRoutingTopic('OMM.fbs')).toBe('/sdn/data/OMM');
    expect(getSchemaRoutingTopic('CDM.fbs')).toBe('/sdn/data/CDM');
  });

  it('should generate peer routing topic', () => {
    expect(getPeerRoutingTopic('QmXyz123')).toBe('/sdn/peer/QmXyz123');
  });
});

File: sdn-server/internal/subscription/subscription_test.go
package subscription

import (
	"encoding/json"
	"testing"
	"time"
)

func TestCreateSubscription(t *testing.T) {
	manager := NewManager()

	config := SubscriptionConfig{
		DataTypes:   []string{"OMM.fbs", "CDM.fbs"},
		SourcePeers: []string{"all"},
		Encrypted:   true,
		Streaming:   true,
		RateLimit:   100,
	}

	sub, err := manager.CreateSubscription(config)
	if err != nil {
		t.Fatalf("Failed to create subscription: %v", err)
	}

	if sub.ID == "" {
		t.Error("Subscription ID should not be empty")
	}

	if sub.Status != StatusActive {
		t.Errorf("Expected status %s, got %s", StatusActive, sub.Status)
	}

	if len(sub.Config.DataTypes) != 2 {
		t.Errorf("Expected 2 data types, got %d", len(sub.Config.DataTypes))
	}
}

func TestCreateSubscriptionValidation(t *testing.T) {
	manager := NewManager()

	tests := []struct {
		name   string
		config SubscriptionConfig
		errMsg string
	}{
		{
			name:   "empty data types",
			config: SubscriptionConfig{SourcePeers: []string{"all"}},
			errMsg: "at least one data type",
		},
		{
			name:   "empty source peers",
			config: SubscriptionConfig{DataTypes: []string{"OMM.fbs"}},
			errMsg: "at least one source peer",
		},
		{
			name: "negative rate limit",
			config: SubscriptionConfig{
				DataTypes:   []string{"OMM.fbs"},
				SourcePeers: []string{"all"},
				RateLimit:   -1,
			},
			errMsg: "rate limit must be non-negative",
		},
		{
			name: "invalid filter operator",
			config: SubscriptionConfig{
				DataTypes:   []string{"OMM.fbs"},
				SourcePeers: []string{"all"},
				Filters:     []QueryFilter{{Field: "test", Operator: "invalid"}},
			},
			errMsg: "invalid operator",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := manager.CreateSubscription(tt.config)
			if err == nil {
				t.Error("Expected error but got none")
			}
		})
	}
}

func TestSubscriptionLifecycle(t *testing.T) {
	manager := NewManager()

	config := SubscriptionConfig{
		DataTypes:   []string{"OMM.fbs"},
		SourcePeers: []string{"all"},
		Encrypted:   true,
		Streaming:   true,
	}

	sub, err := manager.CreateSubscription(config)
	if err != nil {
		t.Fatalf("Failed to create subscription: %v", err)
	}

	// Get subscription
	retrieved, err := manager.GetSubscription(sub.ID)
	if err != nil {
		t.Fatalf("Failed to get subscription: %v", err)
	}
	if retrieved.ID != sub.ID {
		t.Error("Retrieved subscription ID mismatch")
	}

	// Pause subscription
	if err := manager.PauseSubscription(sub.ID); err != nil {
		t.Fatalf("Failed to pause subscription: %v", err)
	}
	retrieved, _ = manager.GetSubscription(sub.ID)
	if retrieved.Status != StatusPaused {
		t.Errorf("Expected status %s, got %s", StatusPaused, retrieved.Status)
	}

	// Resume subscription
	if err := manager.ResumeSubscription(sub.ID); err != nil {
		t.Fatalf("Failed to resume subscription: %v", err)
	}
	retrieved, _ = manager.GetSubscription(sub.ID)
	if retrieved.Status != StatusActive {
		t.Errorf("Expected status %s, got %s", StatusActive, retrieved.Status)
	}

	// Delete subscription
	if err := manager.DeleteSubscription(sub.ID); err != nil {
		t.Fatalf("Failed to delete subscription: %v", err)
	}

	_, err = manager.GetSubscription(sub.ID)
	if err != ErrSubscriptionNotFound {
		t.Error("Expected subscription not found error")
	}
}

func TestProcessMessage(t *testing.T) {
	manager := NewManager()

	config := SubscriptionConfig{
		DataTypes:   []string{"OMM.fbs"},
		SourcePeers: []string{"all"},
		Encrypted:   true,
		Streaming:   true,
	}

	sub, _ := manager.CreateSubscription(config)

	received := make(chan struct{}, 1)
	manager.AddHandler(sub.ID, func(s *Subscription, schema string, data []byte, from string, header *RoutingHeader) {
		received <- struct{}{}
	})

	// Process matching message
	header := NewRoutingHeader("OMM.fbs", "peer1")
	manager.ProcessMessage("OMM.fbs", []byte("{}"), "peer1", header)

	select {
	case <-received:
		// Success
	case <-time.After(time.Second):
		t.Error("Handler was not called")
	}

	// Process non-matching schema
	manager.ProcessMessage("CDM.fbs", []byte("{}"), "peer1", header)
	select {
	case <-received:
		t.Error("Handler should not be called for non-matching schema")
	case <-time.After(100 * time.Millisecond):
		// Expected
	}
}

func TestFilterMatching(t *testing.T) {
	manager := NewManager()

	config := SubscriptionConfig{
		DataTypes:   []string{"OMM.fbs"},
		SourcePeers: []string{"all"},
		Encrypted:   true,
		Filters: []QueryFilter{
			{Field: "OBJECT_NAME", Operator: OpEqual, Value: "ISS"},
		},
	}

	sub, _ := manager.CreateSubscription(config)

	received := make(chan struct{}, 1)
	manager.AddHandler(sub.ID, func(s *Subscription, schema string, data []byte, from string, header *RoutingHeader) {
		received <- struct{}{}
	})

	// Process matching message
	matchingData, _ := json.Marshal(map[string]interface{}{"OBJECT_NAME": "ISS"})
	header := NewRoutingHeader("OMM.fbs", "peer1")
	manager.ProcessMessage("OMM.fbs", matchingData, "peer1", header)

	select {
	case <-received:
		// Success
	case <-time.After(time.Second):
		t.Error("Handler was not called for matching filter")
	}

	// Process non-matching message
	nonMatchingData, _ := json.Marshal(map[string]interface{}{"OBJECT_NAME": "Hubble"})
	manager.ProcessMessage("OMM.fbs", nonMatchingData, "peer1", header)

	select {
	case <-received:
		t.Error("Handler should not be called for non-matching filter")
	case <-time.After(100 * time.Millisecond):
		// Expected
	}
}

func TestRateLimit(t *testing.T) {
	manager := NewManager()

	config := SubscriptionConfig{
		DataTypes:   []string{"OMM.fbs"},
		SourcePeers: []string{"all"},
		Encrypted:   true, // Match the header
		RateLimit:   2,    // Only 2 messages per minute
	}

	sub, _ := manager.CreateSubscription(config)

	// Use a channel to count messages in a thread-safe way
	received := make(chan struct{}, 10)
	manager.AddHandler(sub.ID, func(s *Subscription, schema string, data []byte, from string, header *RoutingHeader) {
		received <- struct{}{}
	})

	header := NewRoutingHeader("OMM.fbs", "peer1")

	// Send 5 messages, only 2 should go through
	for i := 0; i < 5; i++ {
		manager.ProcessMessage("OMM.fbs", []byte("{}"), "peer1", header)
	}

	// Wait for async handlers with timeout
	count := 0
	timeout := time.After(500 * time.Millisecond)
loop:
	for {
		select {
		case <-received:
			count++
		case <-timeout:
			break loop
		}
	}

	if count != 2 {
		t.Errorf("Expected 2 messages (rate limited), got %d", count)
	}
}

func TestSourcePeerFiltering(t *testing.T) {
	manager := NewManager()

	config := SubscriptionConfig{
		DataTypes:   []string{"OMM.fbs"},
		SourcePeers: []string{"peer1", "peer2"},
		Encrypted:   true,
	}

	sub, _ := manager.CreateSubscription(config)

	received := make(chan string, 10)
	manager.AddHandler(sub.ID, func(s *Subscription, schema string, data []byte, from string, header *RoutingHeader) {
		received <- from
	})

	header := NewRoutingHeader("OMM.fbs", "peer1")

	// These should match
	manager.ProcessMessage("OMM.fbs", []byte("{}"), "peer1", header)
	manager.ProcessMessage("OMM.fbs", []byte("{}"), "peer2", header)
	// This should not match
	manager.ProcessMessage("OMM.fbs", []byte("{}"), "peer3", header)

	time.Sleep(100 * time.Millisecond)

	if len(received) != 2 {
		t.Errorf("Expected 2 messages, got %d", len(received))
	}
}

func TestGetRequiredTopics(t *testing.T) {
	manager := NewManager()

	config1 := SubscriptionConfig{
		DataTypes:   []string{"OMM.fbs", "CDM.fbs"},
		SourcePeers: []string{"all"},
	}

	config2 := SubscriptionConfig{
		DataTypes:   []string{"EPM.fbs"},
		SourcePeers: []string{"peer1"},
	}

	manager.CreateSubscription(config1)
	manager.CreateSubscription(config2)

	topics := manager.GetRequiredTopics()

	// Should have topics for OMM, CDM, EPM, and peer1
	expectedTopics := map[string]bool{
		"/sdn/data/OMM":             true,
		"/sdn/data/CDM":             true,
		"/sdn/data/EPM":             true,
		"/spacedatanetwork/sds/OMM.fbs": true,
		"/spacedatanetwork/sds/CDM.fbs": true,
		"/spacedatanetwork/sds/EPM.fbs": true,
		"/sdn/peer/peer1":           true,
	}

	for _, topic := range topics {
		if !expectedTopics[topic] {
			t.Errorf("Unexpected topic: %s", topic)
		}
	}
}

func TestEvaluateFilter(t *testing.T) {
	tests := []struct {
		name     string
		data     map[string]interface{}
		filter   QueryFilter
		expected bool
	}{
		{
			name:     "equal string",
			data:     map[string]interface{}{"name": "ISS"},
			filter:   QueryFilter{Field: "name", Operator: OpEqual, Value: "ISS"},
			expected: true,
		},
		{
			name:     "not equal",
			data:     map[string]interface{}{"name": "ISS"},
			filter:   QueryFilter{Field: "name", Operator: OpNotEqual, Value: "Hubble"},
			expected: true,
		},
		{
			name:     "greater than",
			data:     map[string]interface{}{"altitude": 400.0},
			filter:   QueryFilter{Field: "altitude", Operator: OpGreater, Value: 300.0},
			expected: true,
		},
		{
			name:     "contains",
			data:     map[string]interface{}{"description": "International Space Station"},
			filter:   QueryFilter{Field: "description", Operator: OpContains, Value: "Space"},
			expected: true,
		},
		{
			name:     "nested field",
			data:     map[string]interface{}{"object": map[string]interface{}{"name": "ISS"}},
			filter:   QueryFilter{Field: "object.name", Operator: OpEqual, Value: "ISS"},
			expected: true,
		},
		{
			name:     "in array",
			data:     map[string]interface{}{"type": "satellite"},
			filter:   QueryFilter{Field: "type", Operator: OpIn, Value: []interface{}{"satellite", "debris"}},
			expected: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := evaluateFilter(tt.data, tt.filter)
			if result != tt.expected {
				t.Errorf("Expected %v, got %v", tt.expected, result)
			}
		})
	}
}

func TestNewRoutingHeader(t *testing.T) {
	header := NewRoutingHeader("OMM", "peer123")

	if header.SchemaType != "OMM" {
		t.Errorf("Expected schema type OMM, got %s", header.SchemaType)
	}

	if header.SourcePeer != "peer123" {
		t.Errorf("Expected source peer peer123, got %s", header.SourcePeer)
	}

	if header.TTL != 7 {
		t.Errorf("Expected default TTL 7, got %d", header.TTL)
	}

	if header.Priority != PriorityNormal {
		t.Errorf("Expected normal priority, got %d", header.Priority)
	}

	if !header.Encrypted {
		t.Error("Expected encrypted to be true by default")
	}

	if header.Timestamp == 0 {
		t.Error("Expected timestamp to be set")
	}
}
